torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/embedding.py

```diff
@@ -39,34 +41,25 @@ def sharded_embedding(types, args, kwargs, pg):
     4 GPUs creating 3 shard of (3 x 17) and 1 shard of (1 x 17).
     The algorithm is as follows:
 
-    1. First the input is flattened to 1D and gets sorted so that we can distribute
-       them to the corresponding rank. For example if the given input is
+    1. First the input is broadcasted to all ranks, since this is SPMD we
+       actually do an all_gather for all the inputs resulting in 4 (4 x 6)
+       inputs on each rank. For example if the given input is
        tensor([[6, 5, 2, 9, 6, 3],
                [3, 1, 2, 4, 7, 6],
                [4, 0, 4, 9, 8, 9],
                [8, 6, 6, 4, 6, 1]])
-       Then we have the 1D array like:
-       tensor([6, 5, 2, 9, 6, 3, 3, 1, 2, 4, 7, 6, 4, 0, 4, 9, 8, 9, 8, 6, 6, 4, 6, 1])
-       And sort it:
-       tensor([0, 1, 1, 2, 2, 3, 3, 4, 4, 4, 4, 5, 6, 6, 6, 6, 6, 6, 7, 8, 8, 9, 9, 9])
-       We also record the indices so that we can recover back.
-    2. Next we perform the split by search the index of the chunking
-       boundary. So the above array will be split into 4 parts:
-       tensor([[0, 1, 1, 2, 2], [3, 3, 4, 4, 4, 4, 5],
-              [6, 6, 6, 6, 6, 6, 7, 8, 8], [9, 9, 9])
-       Rearrangement may be needed if the rank order is different from
-       its index in the placement.
-    3. Next, we communicate the length of each part to each rank via all2all
-       so that each rank now knows what input it will get from all other ranks.
-    4. Before we send out the array to other ranks, we need to do the modular operation
-       so that each rank do use that for embedding lookup.
-       The above tensor will look like the below after performing the moduler of 3:
-       tensor([[0, 1, 1, 2, 2], [0, 0, 1, 1, 1, 1, 2],
-              [0, 0, 0, 0, 0, 0, 1, 2, 2], [0, 0, 0])
-    5. Now, each rank receives a matrix (size may vary) and do the lookup. We then use
-       all2all to send the result back to each rank.
-    6. We use the recorded indices to recover the sorted positions and reshape the
-       matrix to (4 x 6 x 17), which is what we need.
+       on rank 0.
+       Then on every rank, we will have this tensor.
+    2. Next, we mask the ID which are not stored on that rank.
+       For example on rank 0, we store ID [0, 1, 2]. We only keep the ID
+       inside the set of numbers. The rest of them will be masked to an extra row.
+       The masked matrix will be used for embedding look up and is like:
+       tensor([[4, 4, 2, 4, 4, 4],
```

Comment:
4 is the masked ID and the reason is that for the weight being sharded there is only 0, 1 and 2. We will add one more dummy entry (4); otherwise, when max_norm is specified, the result is not correct because max_norm will norm the row which is looked up.